keystone: Add fernet as token provider

mjura opened the pull request.

mjura commented:
> Don't you need to synchronize the fernet keys between the cluster nodes in HA mode?

Yes, documentation recommend to use rsync, but I think better would to use some shared storage like drbd. For rsync it would be necessary to sync always tokens after using `keystone-manage fernet_rotate` command. @vuntz What is your suggestion for it?

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -259,6 +259,15 @@
+execute "keystone-manage fernet_setup" do
+  command "keystone-manage fernet_setup \
+  --keystone-user `id -u #{node[:keystone][:user]}` \
+  --keystone-group `id -g #{node[:keystone][:group]}`"
```
You can just use the group name, there is no need for warpping this in id -g

dirkmueller commented on chef/cookbooks/keystone/templates/default/keystone.conf.erb:
```diff
@@ -37,6 +37,11 @@ certfile = <%= @ssl_certfile %>
 ca_certs = <%= @ssl_ca_certs %>
 
+[fernet_tokens]
+
```
extra space

mjura replied:
Should I add this extra space?

dirkmueller commented on crowbar_framework/app/helpers/barclamp/keystone_helper.rb:
```diff
@@ -31,7 +31,8 @@ def server_frontends_for_keystone(selected)
         [
-          ["UUID", "UUID"]
+          ["UUID", "UUID"],
+          ["fernet", "fernet"]
```
User visible Spelling is "Fernet" not "fernet"

mjura replied:
It was done, it is "Fernet"

dirkmueller commented on chef/data_bags/crowbar/template-keystone.json:
```diff
@@ -62,6 +62,10 @@
         "ca_certs": "/etc/keystone/ssl/certs/ca.pem"
       },
+      "fernet": {
+        "key_repository": "/etc/keystone/fernet-keys/",
```
There is no point in making that configurable, its anyway hardcoded in our packaging.

mjura replied:
Ok, I removed this option

dirkmueller commented on chef/data_bags/crowbar/template-keystone.json:
```diff
@@ -57,7 +57,7 @@
       },
       "signing": {
-        "token_format": "UUID",
+        "token_format": "uuid",
```
where is the migration for this?

mjura replied:
It was added

dirkmueller replied:
Thanks. default for new installations should be fernet, could you adjust that?

dirkmueller commented on crowbar_framework/app/helpers/barclamp/keystone_helper.rb:
```diff
@@ -31,7 +31,7 @@ def server_frontends_for_keystone(selected)
       options_for_select(
         [
-          ["UUID", "UUID"],
+          ["uuid", "uuid"],
```
User visible value should be UUID (thats how it is documented upstream, see http://docs.openstack.org/admin-guide/identity-tokens.html)

mjura replied:
It is was changed already, both options for `provider` token are written lowercase
http://docs.openstack.org/developer/keystone/configuration.html#token-provider

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -259,6 +259,15 @@
 end
 
+# Setup a key repository for Fernet tokens
+execute "keystone-manage fernet_setup" do
```
this must only be executed once ! not on every chef run

dirkmueller commented:
@mjura token_rotation is missing (use cron/rsync for that? or a chef databag?)

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -259,6 +259,15 @@
+  --keystone-user `id -u #{node[:keystone][:user]}` \
+  --keystone-group `id -g #{node[:keystone][:group]}`"
+  action :run
+  only_if { node[:keystone][:signing][:token_format] == "fernet" }
```
check for cluster founder missing.

mjura replied:
Done

mjura commented:
It works fine, even with propagating keys to nodes in HA cluster

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -313,6 +313,76 @@
+  end
+
+  # Rotate primary key, which is used for new tokens
+  template "/etc/cron.d/keystone-fernet-rotate" do
```
just put it into cron.hourly then instead of hardcoding a certain time

dirkmueller commented on chef/cookbooks/keystone/recipes/rsync.rb:
```diff
@@ -0,0 +1,37 @@
+# limitations under the License.
+#
+
+directory "/var/run/keystone" do
```
why?

dirkmueller commented on chef/cookbooks/keystone/recipes/rsync.rb:
```diff
@@ -0,0 +1,37 @@
+  )
+end
+
+service "rsync" do
```
not sure if we want to have yet another daemon running..

since you're doing a push anyway, why don't you do that via ssh?

dirkmueller commented on chef/cookbooks/keystone/recipes/rsync.rb:
```diff
@@ -0,0 +1,37 @@
+# Copyright 2011 Dell, Inc.
```
really? :)

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -313,6 +313,76 @@
+  include_recipe "keystone::rsync" if ha_enabled
+
+  # Create directory for key repository
+  directory "/etc/keystone/fernet-keys" do
```
why? fernet_setup below does that

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -313,6 +313,76 @@
+      --keystone-group #{node[:keystone][:group]}"
+    action :run
+    only_if do
+      !node[:keystone][:fernet][:setup] &&
```
File::exist?(/etc/keystone/fernet-keys/0)

dirkmueller commented on chef/cookbooks/keystone/templates/default/rsyncd.conf.erb:
```diff
@@ -0,0 +1,11 @@
+[fernet]
+max connections = 64
+path = /etc/keystone/fernet-keys/
+read only = false
```
so it is writeable for everyone without any protection? that doesn't sound like a very secure solution ;-)

dirkmueller replied:
no, remove, sorry. but I see its consistent with the rest of the file, so thats something for another PR then..

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -313,6 +313,76 @@
+    cluster_nodes.map do |n|
+      next if node.name == n.name
+      node_address = Chef::Recipe::Barclamp::Inventory.get_network_by_type(n, "admin").address
+      rsync_commnad = "rsync /etc/keystone/fernet-keys/* " \
```
no --delete-after ? so we're adding keys without expiring them? is that intended?

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -313,6 +313,76 @@
+  end
+
+  # We want to setup key repository only once
+  ruby_block "mark node for keystone fernet setup" do
```
don't need to save that in node attributes.. the local file is the proof.

dirkmueller commented on chef/cookbooks/keystone/templates/default/keystone.conf.erb:
```diff
@@ -37,6 +37,10 @@ certfile = <%= @ssl_certfile %>
 
+[fernet_tokens]
+
+max_active_keys = <%= node[:keystone][:fernet][:max_active_keys] %>
```
This needs to be calculated based on the token lifetime, see

http://docs.openstack.org/admin-guide/identity-fernet-token-faq.html#i-rotated-keys-and-now-tokens-are-invalidating-early-what-did-i-do

dirkmueller commented on chef/data_bags/crowbar/migrate/keystone/107_add_fernet_tokens_support.rb:
```diff
@@ -0,0 +1,14 @@
+def upgrade(ta, td, a, d)
+  a["fernet"] = {}
+  a["fernet"]["setup"] = ta["fernet"]["setup"]
+  a["fernet"]["max_active_keys"] = ta["fernet"]["max_active_keys"]
```
should not be a separate attribute, needs to be calcluated from the expiration time

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -313,6 +313,76 @@
+      keystone_user: node[:keystone][:user],
+      keystone_command: fernet_rotate_cmd
+    )
+    only_if { !ha_enabled || CrowbarPacemakerHelper.is_cluster_founder?(node) }
```
only on the founder? what if that one is down?

dirkmueller commented:
needs rebase

dirkmueller commented:
https://www.hastexo.com/resources/hints-and-kinks/managing-cron-jobs-pacemaker/ for a hint on how to do this in a cluster

mjura commented:
@dirkmueller : Thank you for all comments, I will included them into next iteration

mjura commented:
I reworked whole setup for propagating Keystone fernet tokens, I still have to create pacemaker resource for crontab

dirkmueller commented on chef/data_bags/crowbar/template-keystone.schema:
```diff
@@ -16,6 +16,7 @@
                     "domain_specific_drivers": { "type": "bool", "required": true },
                     "domain_config_dir": { "type": "str", "required": true },
                     "token_expiration": { "type": "int", "required": true },
+                    "fernet_max_active_keys": { "type": "str", "required": true },
```
1) it shouldn't be a string and
2) it needs to be calculated - see http://docs.openstack.org/admin-guide/identity-fernet-token-faq.html#i-rotated-keys-and-now-tokens-are-invalidating-early-what-did-i-do

mjura replied:
I've changed this implementation https://github.com/crowbar/crowbar-openstack/pull/747/files#diff-28b6adf7bd1f81183dace3201a2720a5R213

dirkmueller commented on chef/cookbooks/keystone/attributes/default.rb:
```diff
@@ -120,3 +120,8 @@
+# Pacemaker Keystone fernet service
+default[:keystone][:ha][:fernet][:agent] = "ocf:heartbeat:symlink"
+default[:keystone][:ha][:fernet][:op][:monitor][:interval] = "10s"
+default[:keystone][:ha][:cron][:agnet] = "lsb:cron"
```
should be using the systemd provider

mjura replied:
On SLES12 cron runs always, I think that this is even not needed

dirkmueller commented on chef/cookbooks/keystone/recipes/fernet_ha.rb:
```diff
@@ -0,0 +1,53 @@
+  params ( {
+    "target" => "/var/lib/keystone/keystone-fernet-rotate",
+    "link" => "/etc/cron.hourly/openstack-keystone-fernet",
+    "backup_suffix" => ".disabled"
```
.disabled is not in the exclude list of /usr/lib/cron/run-crons, so I think you need to choose something else..

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +381,56 @@
+    cluster_nodes.map do |n|
+      next if node.name == n.name
+      node_address = Chef::Recipe::Barclamp::Inventory.get_network_by_type(n, "admin").address
+      rsync_commnad += "rsync -a --delete-after /etc/keystone/fernet-keys " \
```
nothing requires the package to be installed it seems, can you install it?

mjura replied:
I added `package "rsync"`

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +381,56 @@
+    end
+  end
+
+  rsync_commnad = ""
```
can you fix the spelling? this is irritating

mjura replied:
I'm sorry for this, it was fixed

mjura commented:
Enabling tests for this change https://github.com/SUSE-Cloud/automation/pull/1625

mjura commented on chef/cookbooks/keystone/recipes/fernet_ha.rb:
```diff
@@ -0,0 +1,31 @@
+    "target" => "/var/lib/keystone/keystone-fernet-rotate",
+    "link" => "/etc/cron.daily/openstack-keystone-fernet",
+    "backup_suffix" => ".disabled"
+  })
```
I would like to ignore that

mjura commented on chef/cookbooks/keystone/recipes/fernet_ha.rb:
```diff
@@ -0,0 +1,31 @@
+pacemaker_primitive fernet_service_name do
+  agent node[:keystone][:ha][:fernet][:agent]
+  params({
+    "target" => "/var/lib/keystone/keystone-fernet-rotate",
```
I would like to ignore that

mjura commented on chef/cookbooks/keystone/recipes/fernet_ha.rb:
```diff
@@ -0,0 +1,31 @@
+
+pacemaker_primitive fernet_service_name do
+  agent node[:keystone][:ha][:fernet][:agent]
+  params({
```
I would like to ignore that

mjura commented:
@dirkmueller: I've addressed all your comments, it works fine, thank you

dirkmueller commented on chef/cookbooks/keystone/recipes/fernet_ha.rb:
```diff
@@ -0,0 +1,31 @@
+  params({
+    "target" => "/var/lib/keystone/keystone-fernet-rotate",
+    "link" => "/etc/cron.daily/openstack-keystone-fernet",
+    "backup_suffix" => ".disabled"
```
.disabled is not in the list of disabled extensions for cron.daily. can you just delete the file perhaps?

mjura replied:
This will work as follows, RA checks if a file /etc/cron.daily/openstack-keystone-fernet exists, if it does it simply rename it to openstack-keystone-fernet.disabled after the symlink is created. If it does not it directly creates the symlink. During moving this service between nodes in the cluster symlink is removed from inactive nodes. It is fine as it is.

dirkmueller replied:
Right, that's what i said. In case the .disabled file is created, that would get executed then still. Just use a backup suffix that is blacklisted by cron, like for example '~'

mjura commented:
I switched also default value for Keystone token provider to `fernet`.

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -205,6 +205,13 @@
+#       i-rotated-keys-and-now-tokens-are-invalidating-early-what-did-i-do
+# max_active_keys = (token_expiration / rotation_frequency) + 2
+# keystone fernet_rotate job runs daily, it means that rotation_frequency = 24
+max_active_keys = (node[:keystone][:token_expiration] / 24).round + 2
```
token_expiration is in seconds, you're comparing it here in hours, that ain't work

mjura commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +390,65 @@
+      "target" => "/var/lib/keystone/keystone-fernet-rotate",
+      "link" => "/etc/cron.hourly/openstack-keystone-fernet",
+      "backup_suffix" => ".orig"
+    })
```
Ignore

mjura commented:
> key rotation needs to be fixed.. maybe use hourly instead of daily rotation?

You are right `token_expiration` variable from http://docs.openstack.org/admin-guide/identity-fernet-token-faq.html#i-rotated-keys-and-now-tokens-are-invalidating-early-what-did-i-do is not this same `token_expiration` from Keyston chef cookbook. I've changed this to hourly rotation and compute to hours.

> also the ha job doesn't look good

Thank you for improving this. Sorry, accidentally I overwrote your commits.

Fixed

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +390,65 @@
+    execute "propagate fernet keys to all nodes in the cluster" do
+      command rsync_command
+      ignore_failure true
+      action :run
```
so you're running that on every chef run from the founder, independently of who is currently the fernet-token master. that doesn't work. it should only be run once when fernet_setup is run from the founder, or otherwise via the cron job.

mjura replied:
I've changed the logic here, now rsync_command will be executed only once after `keystone-manage fernet_setup` setup.

dirkmueller replied:
yep, looks better now, thanks

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +390,65 @@
+    source "keystone-fernet-rotate.erb"
+    owner "root"
+    group node[:keystone][:group]
+    mode "0755"
```
I guess 0750 might be enough as permissions

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -205,6 +205,14 @@
+# max_active_keys = (token_expiration / rotation_frequency) + 2
+# keystone fernet_rotate job runs hourly, it means that rotation_frequency = 1
+# node[:keystone][:token_expiration] is in seconds and has to be encoded to hours
+max_active_keys = (node[:keystone][:token_expiration] / 3600).round + 2
```
.ceil might be better than .round here, although it doesn't strictly matter since token_expiration is int. I suggest something like:

```ruby
max_active_keys = (node[:keystone][:token_expiration].to_f / 3600).ceil + 2
```

mjura replied:
Done

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +390,64 @@
+    end
+  end
+
+  unless Dir.exist?("/etc/keystone/fernet-keys")
```
might be safer to check for ```File.exist("/etc/keystone/fernet-keys/0") ``` rather than dir (could be empty because $magic went wrong)

mjura replied:
Ok, agree

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +390,64 @@
+    # We would like to propagate fernet keys to all nodes in the cluster
+    execute "propagate fernet keys to all nodes in the cluster" do
+      command rsync_command
+      ignore_failure true
```
I wonder why we want to ignore failures here?

mjura replied:
Would be better to keep it in case if ssh_keys where not propagated properly, but then this will mean that there is something wrong with pacemaker_cluster

mjura replied:
I removed this condition

dirkmueller commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +390,64 @@
+    })
+    op node[:keystone][:ha][:fernet][:op]
+    action :create
+    only_if { CrowbarPacemakerHelper.is_cluster_founder?(node) }
```
maybe we need ha_enabled && ... here. not sure.

mjura commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +389,65 @@
+  pacemaker_primitive "keystone-fernet-rotate" do
+    agent node[:keystone][:ha][:fernet][:agent]
+    params({
+      "target" => "/var/lib/keystone/keystone-fernet-rotate",
```
Ignore

mjura commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +389,65 @@
+
+  pacemaker_primitive "keystone-fernet-rotate" do
+    agent node[:keystone][:ha][:fernet][:agent]
+    params({
```
Ignore

toabctl commented on chef/cookbooks/keystone/recipes/server.rb:
```diff
@@ -381,6 +390,51 @@
+  end
+
+  # Rotate primary key, which is used for new tokens
+  template "/var/lib/keystone/keystone-fernet-rotate" do
```
nit: if we need another iteration, this would be better called ```/var/lib/keystone/keystone-fernet-rotate-and-sync``` because it is doing both, rotating and syncing.

toabctl commented on chef/data_bags/crowbar/template-keystone.json:
```diff
@@ -122,7 +122,7 @@
       "crowbar-revision": 0,
       "crowbar-applied": false,
-      "schema-revision": 106,
+      "schema-revision": 108,
```
what about migration number 107?

mjura replied:
But we have two migration scripts in two different commits. I can merge this but I would have to also squash commits, which cause then we will lose some history of changes.

toabctl replied:
Ah. Sorry. Missed that.
